### PR TITLE
0.4.0

### LINKED ISSUE #12: `lean` leaks
As [discovered by Miri](https://github.com/maciejhirsz/beef/pull/11#discussion_r394348721), the `lean` variant of beef leaks memory. I verified that by adding this patch
```diff
diff --git a/src/generic.rs b/src/generic.rs
index add1a9b..9fa0981 100644
--- a/src/generic.rs
+++ b/src/generic.rs
@@ -170,6 +170,7 @@ where
     #[inline]
     fn drop(&mut self) {
         if let Some(capacity) = self.capacity() {
+            panic!("dropping!");
             unsafe { T::owned_from_parts::<U>(self.inner, capacity) };
         }
     }
```
and then running `cargo test lean`. Clearly that should drop the contents and thus panic at some point, but it does not.

### PR DESCRIPTION
**Pull request by maciejhirsz**
+ Fixed a memory leak in `beef::lean::Cow`, found thanks to Miri and @RalfJung (closes #12).
+ Added `impl_serde` feature which provides a `Serialize` and `Deserialize` trait implementations. ~Note that it's currently impossible to make it work with `#[serde(borrow)]` ([relevant PR](https://github.com/serde-rs/serde/pull/1754)).~
+ Added `is_borrowed` and `is_owned` methods.
+ `beef::Cow<'a, T>` is now a type alias to `beef::generic::Cow<'a, T, Wide>` where `Wide` is private. This doesn't change anything, merely hides `Option<NonZeroUsize>` as implementation detail and makes errors more readable.
+ Due to internal refactoring, `const_fn` feature now adds two separate functions instead of one: `const_str` and `const_slice` for borrowing `&'static str` and `&'static [T]` respectively. This is the only breaking change, this release.

### CODE REVIEW

**Comment by RalfJung** [comment]
Looks like one of the new tests does a lot, and since Miri is pretty slow that just takes waaaay too long.

What we do in other test suites is to either disable that test in Miri, or (better) reduce the number of iterations in Miri. You can use `cfg(miri)` to test if this is being compiled for Miri.

**Comment by maciejhirsz** [comment]
Great, thank you a lot for the help @RalfJung! Since Miri was complaining about the fat pointer, and since according to documentation of `ptr::slice_from_raw_parts` creating a slice with invalid length is UB (though I don't think there was any unsafety in practice), I've split the pointer, new Cow looks like:

```rust
pub struct Cow<'a, T: Beef + ?Sized + 'a, U: Capacity> {
    /// Pointer to data
    ptr: NonNull<T::PointerT>,

    /// This usize contains length, but it may contain other
    /// information pending on impl of `Capacity`, and must therefore
    /// always go through `U::len` or `U::unpack`
    fat: usize,

    /// Capacity field. For `beef::lean::Cow` this is 0-sized!
    cap: U::Field,

    /// Lifetime marker
    marker: PhantomData<&'a T>,
}
```

I've renamed `len` to `fat` where appropriate to differentiate between usable length from which a fat pointer can be made, and potentially invalid length that might contain capacity as is the case in owned `lean::Cow`. I was afraid I wouldn't be able to make `const_fn` borrows work without a fat pointer, but it turns out to be workable with separate functions for str and slices.

File: src/lib.rs [hunk 1]
```diff
@@ -164,6 +164,7 @@ macro_rules! test { ($tmod:ident => $cow:path) => {
         }
 
         #[test]
+        #[cfg_attr(miri, ignore)]
```
Comment (RalfJung) [inline src/lib.rs]: The more elegant alternative would be sth like
```rust
#[cfg(not(miri))]
let iterations = 1024;
#[cfg(miri)]
let iterations = 10;

for i in 0..iterations {
```

**Comment by CAD97** [comment]
> since according to documentation of `ptr::slice_from_raw_parts` creating a slice with invalid length is UB

That's not my reading. And `ptr::slice_from_raw_parts` is safe to call, so it cannot be the case.

> This function is safe, but actually using the return value is unsafe.

What's unsafe is then dereferencing the constructed raw slice pointer, which is UB via all the normal pointer provenance rules.

**Comment by maciejhirsz** [comment]
> What's unsafe is then dereferencing the constructed raw slice pointer, which is UB via all the normal pointer provenance rules.

Right, the problem is I _was_ dereferencing this invalid pointer, even though all I was doing with it was calling `.len()`. Since there is no `ptr::slice_into_raw_parts` I'd have to do something ugly like transmute it into `(usize, usize)` and read that, but that too cursed for me.

**Comment by maciejhirsz** [comment]
That said, these are equivalent apparently:

```rust
#[no_mangle]
fn len(ptr: *const [u8]) -> usize {
    unsafe { std::mem::transmute::<_, (usize, usize)>(ptr).1 }
}

#[no_mangle]
fn len_UB(ptr: *const [u8]) -> usize {
    unsafe { (*ptr).len() }
}
```

```asm
len:
	movq	%rsi, %rax
	retq
.set len_UB, len
```

**Comment by CAD97** [comment]
I think I've mentioned it before, but you actually can get the length from a raw slice pointer on stable today:

```rust
fn raw_slice_len<T>(slice: ptr::NonNull<[T]>) -> usize {
    let slice = slice.as_raw() as *mut [()];
    (&*slice).len()
}
```

This works because `[()]` is a zero sized type with minimal alignment, and so long as the pointer is properly aligned and nonnull, Rust says dereferencing a pointer to a zero-sized place is defined and in fact does not even depend on the pointer having been allocated.

_That_ said, I think the split design for `beef::Cow` is probably better in the long run.

**Comment by RalfJung** [comment]
Cc https://github.com/rust-lang/rust/issues/60639

**Comment by RalfJung** [comment]
@maciejhirsz 
> That said, these are equivalent apparently

It is meaningless to look at the assembly of a program that has UB. In that case the compiler just outputs *something* which might or might not be valid assembly, and you can read it like tea leafs. Also see [this blog post](https://www.ralfj.de/blog/2019/07/14/uninit.html).

**Comment by maciejhirsz** [comment]
> _That_ said, I think the split design for `beef::Cow` is probably better in the long run.

Yeah. I'm just running some benchmarks against my experimental json-rust branch and there are no performance drops I can measure, so I'm quite happy with this.

Just waiting for serde 1.0.105 to be publish now so I can add tests for borrows from json and going to merge and publish this (speaking of, those probably should be disabled in Miri).

**Comment by RalfJung** [comment]
> speaking of, those probably should be disabled in Miri

Why that?

**Comment by maciejhirsz** [comment]
I didn't think we need to run `serde_json` parsing through Miri, and Serialize / Deserialize impls are both one-liners. Looking at travis logs now, it only adds a second, somehow I thought it would make a bigger difference.

**Comment by RalfJung** [comment]
Miri is slow but not *that* slow. ;)

**Pull request merged**